### PR TITLE
Teach typedConstructorPattern to ignore non-redefining unapplies

### LINKED ISSUE #12250: Case classes with additional custom unapply produce pattern match warning
## reproduction steps

using Scala 2.13.4, I have a case class with an additional custom unapply method. 

```scala
final case class Foo(value: String)

object Foo {
  def unapply(str: String): Option[Foo] = Some(Foo(str))

  def extract(id: Foo): String =
    id match {
      case Foo(a) => a
    }
}
```

## problem

Gives a warning: 

```
match may not be exhaustive.
[warn] It would fail on the following input: Foo(_)
[warn]     id match {
[warn]     ^
```

Adding a further unapply method as suggested here #12232 did make the warning go away: 

```scala
def unapply(arg: Foo): Some[String] = Some(arg.value)
```

However, the discussion on that issue seemed to point to not using case classes as the root of the problem. It seems like a bit of a gotcha that adding another unapply method breaks the out-of-the-box pattern matching for case classes.

### PR DESCRIPTION
**Pull request by dwijnand**
In pos/t12250.scala Foo defines an unapply which doesn't redefine the
synthetic one.  Thus unapply is overloaded.  In the pattern being typed,
the synthetic one is indeed the one seeked, so despite being overloaded,
we should convertToCaseConstructor.

Fixes https://github.com/scala/bug/issues/12250

### CODE REVIEW

**Comment by lrytz** [comment]
🤔
```
$> cat A.scala
final case class Foo(value: String)

object Foo {
  def unapply(str: String): Option[Foo] = Some(Foo(str))

  def extract(id: Foo): String =
    id match {
      case Foo(a) => a
    }
}
$> scpr 9462 A.scala
A.scala:7: warning: match may not be exhaustive.
It would fail on the following input: Foo(_)
    id match {
    ^
1 warning
```

**Comment by dwijnand** [comment]
Oh, I took out the -Werror I had initially... because I somehow convinced myself pos tests fail on warnings... 🤯

**Comment by dwijnand** [comment]
Fixed my boolean algebra mistake.

File: src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala [hunk 1]
```diff
@@ -97,9 +97,9 @@ trait PatternTypers {
       // Dueling test cases: pos/overloaded-unapply.scala, run/case-class-23.scala, pos/t5022.scala
       // A case class with 23+ params has no unapply method.
       // A case class constructor may be overloaded with unapply methods in the companion.
-      // A case class maybe have its own custom unapply (so non-synthetic) scala/bug#11252
-      // Unapply methods aren't `isCaseApplyOrUnapply` in Scala 3 tasty/run/src-2/tastytest/TestColour.scala
-      else if (canElide && caseClass.isCase && !member.isOverloaded && (member == NoSymbol || member.isSynthetic))
+      // A case class may have its own custom unapply (so non-synthetic) pos/t11252.scala
+      // A case class may have its own custom unapply unrelated to deconstructing pos/t12250.scala
+      else if (canElide && caseClass.isCase && (!member.isOverloaded && (member == NoSymbol || member.isSynthetic) || member.alternatives.exists(_.isSynthetic) && caseClass.tpe =:= pt))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala]: I don't know my way around here that well; IIUC this assumes that the synthetic `unapply` overload is the one selected - could it be that overloading resolution picks a different one, would it be fine to go in`convertToCaseConstructor`?

Comment (dwijnand) [reply src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala]: I think there is only ever one synthetic unapply for a case class (`caseClass.isCase`).  And your understanding is correct, except it's not really "selected": the presence of the case class unapply means that using its constructor is the right course of action.

Now that I think about it, I should review the case-class-23 situation.  Basically it never had an unapply and we rewrote it to use its constructor.  What happens in the various situations of redefining its unapply or defining an un-related unapply? todo += 1

Comment (dwijnand) [reply src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala]: OK, the situation is there's an issue (IMO) but thankfully my change here didn't affected it.

Comment (dwijnand) [inline src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala]: Talking about this diff in passing @adriaanm mentioned `isApplicable` in Infer, which is already `import infer._`-ed above, so I should see if I can use that over `=:=`.

Comment (dwijnand) [reply src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala]: Turned out not to be necessary. `=:=` is right, and I wrote the super-type/sub-type variations to prove that it's right (or at least convinced myself it's right... confirmation bias...).

I also reviewed how user-defined unapplies (and applies) are determined (#5816) and it's using Type#matchesType.  But, again, `=:=` looks to be correct while `<:<` isn't, which was my original thought.

**Comment by dwijnand** [comment]
Seth was ready to merge, but I wanted to review this detail, so it can be merged now once it's green.

**Pull request merged**